### ISSUE TITLE
Make parametrized destination a class with constructor

### ISSUE THREAD

**Issue by hrach** [raamcosta/compose-destinations#16]
Currently, adding parameters adds a `withArgs()` method. This method has the parameters nicely typed, but the destination spec stays an `object`. This brings possible runtime issues. E.g. adding a param to param-less destination will crash in runtime.

Please change the syntax of such objects to become classes and parameters will be the contructor parameters.

This:
```kotlin
object MyScreenDestination : Destination {
    override val route = "my_screen/{number}"
    override val arguments = listOf(
        navArgument("number") {
            type = NavType.IntType
            nullable = false
        }
    )
    @Composable
    override fun Content(
        navController: NavHostController,
        navBackStackEntry: NavBackStackEntry,
        situationalParameters: Map<Class<*>, Any>,
    ) {
        // ...
    }
    fun withArgs(
        number: Int,
    ): String { /* ... */ }
}
```
will become this:
```kotlin
class MyScreenDestination(
     number: Int,
) : Destination {
    override val route = "my_screen/{number}"
        .replace("{number}", number.toString())
    override val arguments = listOf(
        navArgument("number") {
            type = NavType.IntType
            nullable = false
        }
    )
    @Composable
    override fun Content(
        navController: NavHostController,
        navBackStackEntry: NavBackStackEntry,
        situationalParameters: Map<Class<*>, Any>,
    ) {
        // ...
    }
}
```

**Comment by raamcosta** [comment]
Hi @hrach !

So this is a very interesting idea! 👏 
I have toyed around in my head with something similar, but haven't had time to actually fiddle around in the code.

I do see some issues with your proposal, but at the same time, it makes it clear that there are two concepts here:

1. Destination that is used to fill the navigation graph when `DestinationsNavHost` is called. That one needs to "unresolved" route, the list of arguments, deep links, and the Content method.
2. The class you want to instantiate which only needs the "resolved" route but nothing else.

Besides, I see a flaw of the current design which is that my navigator accepts an instance of `Routed` but it needs to be a "resolved" route. And in these Destinations it may not be if the user calls the navigate method like `navigator.navigate(DestinationThatHasArguments)` instead of `navigator.navigate(DestinationThatHasArguments.withArgs([ARGUMENTS])`.

In conclusion, I definitely want to improve this. I will post my solution here before going ahead with it so that you can give me your valuable feedback 🙂

**Comment by hrach** [comment]
>it makes it clear that there are two concepts her

ah, of course! I forget about this :X Yes, probably not so easy as I've imagined :)

**Comment by raamcosta** [comment]
My initial draft:

```kotlin
package com.ramcosta.composedestinations

import androidx.compose.runtime.Composable
import androidx.navigation.NavBackStackEntry
import androidx.navigation.NavHostController
import androidx.navigation.NavType
import androidx.navigation.navArgument
import com.ramcosta.composedestinations.navigation.DependenciesContainerBuilder
import com.ramcosta.composedestinations.navigation.DestinationDependenciesContainer

import androidx.compose.animation.AnimatedVisibilityScope
import androidx.compose.animation.ExperimentalAnimationApi
import androidx.lifecycle.SavedStateHandle
import androidx.navigation.navDeepLink
import com.ramcosta.composedestinations.spec.Routed
import com.ramcosta.samples.destinationstodosample.destinations.profile.DEFAULT_GROUP
import com.ramcosta.samples.destinationstodosample.destinations.profile.DEFAULT_ID
import com.ramcosta.samples.destinationstodosample.destinations.profile.ProfileScreen
import com.ramcosta.samples.destinationstodosample.destinations.profile.ProfileScreenNavArgs
import com.ramcosta.samples.destinationstodosample.destinations.profile.ProfileTransitions
import com.ramcosta.samples.destinationstodosample.destinations.profile.ProfileUiEvents
import com.ramcosta.samples.destinationstodosample.destinations.profile.ProfileUiState

class ProfileScreenDestination(
	id: Long = DEFAULT_ID,
	groupName: String? = DEFAULT_GROUP,
): ParameterizedRouted {

	override val parameterizedRoute = buildRoute(id, groupName)

	private fun buildRoute(
		id: Long,
		groupName: String?,
	): String {
		var route = Spec.route
		if (groupName != null) {
			route = route.replace("{groupName}", groupName)
		}

		return route
			.replace("{id}", id.toString())
	}

	companion object Spec : Destination {

		override val route = "profile_screen?id={id}?groupName={groupName}"

		override val arguments get() = listOf(
			navArgument("id") {
				type = NavType.LongType
				defaultValue = DEFAULT_ID
			},
			navArgument("groupName") {
				type = NavType.StringType
				nullable = true
				defaultValue = DEFAULT_GROUP
			}
		)

		override val deepLinks get() = listOf(
			navDeepLink {
				uriPattern = "https://destinationssample.com/profile_screen?id={id}?groupName={groupName}"
			}
		)

		override val style = ProfileTransitions

		@Composable
		override fun Content(
			navController: NavHostController,
			navBackStackEntry: NavBackStackEntry,
			dependenciesContainerBuilder: @Composable DependenciesContainerBuilder.() -> Unit
		) {
			val container = DestinationDependenciesContainer().apply { dependenciesContainerBuilder() }
			val animatedVisibilityScope = container.get<AnimatedVisibilityScope>()
			animatedVisibilityScope.ProfileScreen(
				uiState = container.get<ProfileUiState>(),
				uiEvents = container.get<ProfileUiEvents>()
			)
		}
	}
}
```

I would do this with all Destinations though, not just the ones with arguments. I think that would be a more consistent API.

**Comment by raamcosta** [comment]
Second approach, simply replacing withArgs with `operator fun invoke` and returning a Routed object. Also, make it so that `DestinationSpec` is no longer `Routed`. This way, `Routed` is only used for objects with a route ready to be used in navigation.

```kotlin
	operator fun invoke(
		id: Long = DEFAULT_ID,
		groupName: String? = DEFAULT_GROUP,
	): Routed {
		var route = route
		if (groupName != null) {
			route = route.replace("{groupName}", groupName)
		}

		return object: Routed {
			override val route = route.replace("{id}", id.toString())
		}
	}

```

**Comment by raamcosta** [comment]
I went ahead with the second approach. Will be releasing a new version with it today.
Let me know if it works well for your use cases, if so, close the issue.

**Comment by hrach** [comment]
Thanks, will check it out. TBH I've just started with the library and those issue were my first feeling/quick feedback to help you to iterate :)

**Comment by raamcosta** [comment]
New version is out and ready in maven central.

This has been great feedback @hrach! Please keep it up, let's make navigation great :D

**Comment by hrach** [comment]
Great, works nicely. There is a minimal difference - in the class solution sees the constructor right ahead (when going to definition). With object it's not so obvious (but at least the invoke is on the top 🎉)

I'd guess we may close it.

**Comment by raamcosta** [comment]
Yes, that was on purpose (it being on the top). 
Glad it works for you 🙂
